### PR TITLE
Fix #63: Add an Express Web Server

### LINKED ISSUE #63: Create HTTP server
We need to create a web server for telescope. We should add what has been mentioned in #62 to this.

### PR DESCRIPTION
**Pull request by miggs125**
fixes #63. This adds nodemon for debug builds as well as a started web server file with a health check. We can discuss additional endpoints here and I can add them in as we discuss them. I've added minimist so that the server can be started on a different port if needed. This can be used like so:  

```bash
 node server.js --port=4040  # starts server at port 4040 
```

### CODE REVIEW

**Comment by miggs125** [comment]
Oh wow I totally missed that. Will fix tonight!

File: package.json [hunk 1]
```diff
@@ -23,14 +24,19 @@
     "bent": "^7.0.2",
     "bull": "^3.11.0",
     "dotenv": "^8.2.0",
+    "express": "^4.17.1",
+    "express-healthcheck": "^0.1.0",
     "feedparser-promised": "^2.0.1",
+    "minimist": "^1.2.0",
```
Comment (humphd) [inline package.json]: Let's remove `minimist` and `opn`.  The latter isn't used at all, and the former isn't needed, since we won't be starting the server directly from the command line (see comments below).

File: package.json [hunk 2]
```diff
@@ -11,7 +11,8 @@
     "pretest": "npm run lint",
     "test": "npm run jest",
     "jest": "jest",
-    "start": "node src"
+    "start": "node src",
+    "debug": "nodemon server.js"
```
Comment (humphd) [inline package.json]: I think `debug` is too vague, since we are potentially going to have multiple entry points/apps running at the same time.  Should this be `web` and `web-debug` or `server` and `server-debug`?

Also, you've removed `src/index.js` (by accident I hope?), and we should think about integrating the code in there with the web server startup.  See my comment below on this.

File: server.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+require('./src/config');
+const express = require('express');
+const argv = require('minimist');
+const path = require('path');
+const healthcheck = require('express-healthcheck');
+
+const args = argv(process.argv.slice(2));
+const HTTP_PORT = process.env.PORT || args.port || 3000;
```
Comment (humphd) [inline server.js]: Let's not allow passing args in like this. We have an `env` system in place, let's just use that.  If people want to override it on the command line, they can with `PORT=12345 npm start`.

File: server.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+const server = express();
```
Comment (humphd) [inline server.js]: Let's follow best practices and break the definition of express app away from the web server, see https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/projectstructre/separateexpress.md.  This will make testing a lot easier.

File: server.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+server.use('/health', healthcheck());
+
+// *********************  ROUTES  *****************************
+
+server.get('/', (req, res) => {
+  res.sendFile(path.join(__dirname, 'index.html'));
+});
+
+server.listen(HTTP_PORT, () => {
+  console.log(`Telescope listening on port ${HTTP_PORT}`);
```
Comment (humphd) [inline server.js]: When #35 lands, we can switch to use proper logging here.  cc @hansal7014

File: src/index.js [hunk 6]
```diff
@@ -1,52 +0,0 @@
-/**
```
Comment (humphd) [inline src/index.js]: Please restore this file.

File: server.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+const healthcheck = require('express-healthcheck');
+
+const args = argv(process.argv.slice(2));
+const HTTP_PORT = process.env.PORT || args.port || 3000;
+const server = express();
+
+
+// ********************* MIDDLEWARE **************************
+
+server.use('/health', healthcheck());
```
Comment (humphd) [inline server.js]: This would be pretty easy to write a test for, see https://github.com/goldbergyoni/nodebestpractices/blob/master/sections/projectstructre/separateexpress.md#example-test-your-api-in-process-using-supertest-popular-testing-package

File: server.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+// *********************  ROUTES  *****************************
+
+server.get('/', (req, res) => {
```
Comment (humphd) [inline server.js]: This should also have a basic test

**Comment by Grommers00** [comment]
any update on this?

**Comment by miggs125** [comment]
@Grommers00 I updated changes to reflect the ongoing discussion. Waiting for reviews.

File: package.json (hunk 2 above)
Comment (humphd) [reply package.json]: This is still outstanding...

File: server.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+const app = require('./app.js');
+
+const HTTP_PORT = process.env.PORT || 3000;
+
+const server = app.listen(HTTP_PORT, () => {
+  console.log(`Telescope listening on port ${HTTP_PORT}`);
+});
+
+process.on('unhandledRejection', (err) => {
+  console.log('UNHANDLED REJECTION:  Shutting down...');
```
Comment (humphd) [inline server.js]: Once #142 lands, we can move all this into proper logs.  For now, this is good.

File: app.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+const healthcheck = require('express-healthcheck');
+
+const router = require('./src/routes/index');
+
+const app = express();
+
+app.use('/health', healthcheck());
```
Comment (humphd) [inline app.js]: Question: why here and not in your router?

File: app.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+const healthcheck = require('express-healthcheck');
+
+const router = require('./src/routes/index');
```
Comment (humphd) [inline app.js]: ```const router = require('./src/routes');``` (no need for `index`

File: package.json [hunk 12]
```diff
@@ -31,7 +34,15 @@
     "eslint-config-airbnb-base": "^14.0.0",
     "eslint-plugin-import": "^2.18.2",
     "jest": "^24.9.0",
+    "nodemon": "^1.19.4",
     "stylelint": "^11.1.1",
-    "stylelint-config-recommended": "^3.0.0"
+    "stylelint-config-recommended": "^3.0.0",
+    "supertest": "^4.0.2"
+  },
+  "jest": {
```
Comment (humphd) [inline package.json]: This seems unrelated to your changes.  We have an issue on #120, can you comment there, and they can add this?

File: test/server.test.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+const request = require('supertest');
+const app = require('../app');
+
+describe('Health Check', () => {
+  it('should return a JSON object with property "uptime"', async () => {
+    const res = await request(app).get('/health');
+    expect(res.statusCode).toEqual(200);
+    expect(res.body).toHaveProperty('uptime');
+  });
+});
```
Comment (humphd) [inline test/server.test.js]: Let's add one more test that `/` returns a 200 and a content-type of text/html...

File: app.js (hunk 10 above)
Comment (miggs125) [reply app.js]: I added it here because it is middleware so I figured it belonged in app.js but it also makes sense to have it in the router. If it makes more sense to have it in router I will move it there.

File: package.json (hunk 12 above)
Comment (miggs125) [reply package.json]: yup! I'll remove it from here and comment in #120

**Comment by miggs125** [comment]
@humphd I've made the changes requested. I will make changes to where the /health middleware will be included pending the outcome of our /health middlware [conversation](https://github.com/Seneca-CDOT/telescope/pull/141#discussion_r345855477). Once this is resolved, I will rebase all of the commits into something clearer and comprehensive commit message.
